interfaces/builtin/usb_gadget.go

```diff
@@ -29,25 +48,199 @@ const usbGadgetBaseDeclarationSlots = `
+	}
+
+	return nil
+}
+
+func ffsMounts(plug interfaces.Attrer) ([]map[string]any, error) {
+	var mounts []map[string]any
+	err := plug.Attr("ffs-mounts", &mounts)
+	if err != nil && !errors.Is(err, snap.AttributeNotFoundError{}) {
+		return nil, mountAttrTypeError
+	}
+	return mounts, nil
+}
+
+func enumerateFFSMounts(mounts []map[string]any, fn func(mi *ffsMountInfo) error) error {
+	for _, mount := range mounts {
+		name, ok := mount["name"].(string)
+		if !ok {
+			return fmt.Errorf(`usb-gadget FunctionFS mount "name" must be a string`)
+		}
+
+		where, ok := mount["where"].(string)
+		if !ok {
+			return fmt.Errorf(`usb-gadget FunctionFS mount "where" must be a string`)
+		}
+
+		persistent := false
+		persistentValue, ok := mount["persistent"]
+		if ok {
+			if persistent, ok = persistentValue.(bool); !ok {
+				return fmt.Errorf(`usb-gadget FunctionFS mount "persistent" must be a boolean`)
+			}
+		}
+
+		mountInfo := &ffsMountInfo{
+			name:       name,
+			where:      where,
+			persistent: persistent,
+		}
+
+		if err := fn(mountInfo); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (iface *usbGadgetInterface) BeforeConnectPlug(plug *interfaces.ConnectedPlug) error {
+	// The systemd.ListMountUnits() method works by issuing the command
+	// "systemctl show *.mount", but globbing was only added in systemd v209.
+	if err := systemd.EnsureAtLeast(209); err != nil {
+		return err
+	}
+
+	mounts, err := ffsMounts(plug)
+	if err != nil {
+		return err
+	}
+	return enumerateFFSMounts(mounts, func(mi *ffsMountInfo) error {
+		return mi.validate()
+	})
+}
+
+func (iface *usbGadgetInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	usbGadgetSnippet := bytes.NewBuffer(nil)
+	emit := func(f string, args ...any) {
+		fmt.Fprintf(usbGadgetSnippet, f, args...)
+	}
+	snapInfo := plug.Snap()
+
+	// emit rules required generally for the usb-gadget interface
+	emit(`
+  # https://www.kernel.org/doc/Documentation/usb/gadget_configfs.txt
+  # Allow creating new gadgets under usb_gadget, which is creating
+  # new directories
+  /sys/kernel/config/usb_gadget/ rw,
+  # Allow creating sub-directories, symlinks and files under those
+  # directories
+  /sys/kernel/config/usb_gadget/** rw,
+
+  # Allow access to UDC
+  /sys/class/udc/ r,
+`)
+
+	mounts, err := ffsMounts(plug)
+	if err != nil {
+		return err
+	}
+	if len(mounts) > 0 {
+		// add required rules necessary to actually perform the ffs mounts
+		emit(`
+  # Rules added by the usb-gadget interface
+  # due to ffs mount-support
+  capability sys_admin,  # for mount
+
+  owner @{PROC}/@{pid}/mounts r,
+  owner @{PROC}/@{pid}/mountinfo r,
+  owner @{PROC}/self/mountinfo r,
+
+  /{,usr/}bin/mount ixr,
+  /{,usr/}bin/umount ixr,
+  # mount/umount (via libmount) track some mount info in these files
+  /run/mount/utab* wrlk,
```

Comment:
This is wrong as libmount would update a file that is describing the reality of a mount namespace of the snap but /run/ is actually from the host.

We have the same problem in more places. For now I'd quiet deny this instead.

Reply:
I was not aware of this - this snippet is taken from `mount-control`

Reply:
Thanks, might be worth looking at mount-control. There the consequence is not quite clear-cut as depending on location it might or might not be correct.